Scripting: add support for fine-grained settings

javanna opened the pull request:
Note that this PR is against `1.x`.

Allow to enable/disable scripting based on their source (where they get loaded from), the  operation that executes them and their language.

The settings cover the following combinations:

- mode: `on`, `off`, `sandbox`
- source: `indexed`, `dynamic`, `file`
- engine: `groovy`, `expression`, `mustache`, etc
- operation: `update`, `search`, `aggs`, `mapping`

The following settings are supported for every engine:

```
script.engine.groovy.indexed.update:    sandbox/on/off
script.engine.groovy.indexed.search:    sandbox/on/off
script.engine.groovy.indexed.aggs:      sandbox/on/off
script.engine.groovy.indexed.mapping:   sandbox/on/off
script.engine.groovy.dynamic.update:    sandbox/on/off
script.engine.groovy.dynamic.search:    sandbox/on/off
script.engine.groovy.dynamic.aggs:      sandbox/on/off
script.engine.groovy.dynamic.mapping:   sandbox/on/off
script.engine.groovy.file.update:       sandbox/on/off
script.engine.groovy.file.search:       sandbox/on/off
script.engine.groovy.file.aggs:         sandbox/on/off
script.engine.groovy.file.mapping:      sandbox/on/off
```

For ease of use, the following more generic settings are supported too:

```
script.indexed: sandbox/on/off
script.dynamic: sandbox/on/off
script.file:    sandbox/on/off
```
and
```
script.update:  sandbox/on/off
script.search:  sandbox/on/off
script.aggs:    sandbox/on/off
script.mapping: sandbox/on/off
```
These will be used to calculate the more specific settings, using the stricter setting of each combination. Operation based settings have precedence over conflicting source based ones.

Note that the `mustache` engine is affected by generic settings applied to any language, while native scripts aren't as they are static by definition.

Also, the previous `script.disable_dynamic` setting can now be deprecated.

Closes #6418 
Closes #10274

javanna commented:
@s1monw can you have a look please? @clintongormley can you please review the updated docs and the new settings api-wise?

@dadoonet have time to double check that this breaks only potentially rivers, similar to #9992 ?

s1monw commented on src/main/java/org/elasticsearch/script/ScriptMode.java:
```diff
@@ -0,0 +1,62 @@
+
+    static ScriptMode parse(String input) {
+        input = input.toLowerCase(Locale.ROOT);
+        switch(input) {
```
do we have to have a gazillion different options here? can we just use `valueOf(input.toUpperCase())`

javanna replied:
we can but we usually tend to be "understanding" with user input. I don't think this is a big problem, I actually did it because the suggested syntax ("enable" and "disable") differs from what we do in other places. Also whenever we accept a boolean we accept `on`, `off`, `true` and `false` if I remember correctly

s1monw replied:
then use `org.elasticsearch.common.Booleans` in these cases. This one has 3 option can't we just have the 3 options and be done with it? I don't get why we need to complicate things with a gazillion options

javanna replied:
sounds good to me I am going to use `Booleans`, much better. @clintongormley that means I am going to drop support for `enable` and `disable` as values, as we never supported those anywhere else. We support `on`,`true`,`1`,`yes`,`off`,`false`,`0`,`no` or `sandbox`. I will update the docs too

s1monw commented on src/main/java/org/elasticsearch/script/ScriptModes.java:
```diff
@@ -0,0 +1,202 @@
+import java.util.Set;
+import java.util.TreeMap;
+
+public class ScriptModes {
```
can we have a javadoc comment here?

s1monw replied:
I really like this class since it's so self-contained and has all the tests etc. no weird guice bloat etc. NICE!

javanna replied:
cool glad to hear, I like it too ;) ... adding javadocs

s1monw commented on src/main/java/org/elasticsearch/script/ScriptModes.java:
```diff
@@ -0,0 +1,202 @@
+
+public class ScriptModes {
+
+    static final String SCRIPT_SETTINGS_PREFIX = "script.";
```
can those be private constants?

javanna replied:
using them in tests

s1monw commented on src/main/java/org/elasticsearch/script/ScriptModes.java:
```diff
@@ -0,0 +1,202 @@
+        }
+
+        //read deprecated disable_dynamic setting, apply only if none of the new settings is used
+        String disableDynamicSetting = settings.get(ScriptService.DISABLE_DYNAMIC_SCRIPTING_SETTING);
```
maybe we can break this up into several methods and test them separately as well?

javanna replied:
I broke it up in different methods that process their own part fo the settings in order. That said I am not sure about testing. I think how we currently test this class is pretty good and testing these methods separately would create some duplication, not sure it's really worth it. At the end of the day the coverage here is already high and the amount of test code is quite big too.

s1monw commented on src/main/java/org/elasticsearch/script/ScriptModes.java:
```diff
@@ -0,0 +1,202 @@
+    final ImmutableMap<String, ScriptMode> scriptModes;
+
+    ScriptModes(Map<String, ScriptEngineService> scriptEngines, Settings settings) {
+        this.logger = Loggers.getLogger(getClass(), settings);
```
I think we should rather pass the logger to it or don't log here at all. we only use it for this:

```Java
 logger.warn("ignoring [disable_dynamic] setting value as conflicting scripting settings have been specified");
```

IMO we should rather collect the conflicting settings and provide it as a list and then log in the `ScriptService` with all the settings that are conflicting...

javanna replied:
done

s1monw commented on src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -325,7 +329,10 @@ private String validateScriptLanguage(String scriptLang) {
-    private String getScriptFromIndex(Client client, String scriptLang, String id) {
+    String getScriptFromIndex(String scriptLang, String id) {
+        if (client == null) {
+            throw new ElasticsearchIllegalArgumentException("Got an indexed script with no Client registered.");
```
this should be an IllegalStateException IMO?

javanna replied:
grrrr I didn't change this.... it's always been this way and to be honest I don't fully follow this whole optional client thing, and throwing an exception only from getScriptFromIndex but not from the other methods that use the client.... can we maybe clean this up in a separate PR?

s1monw replied:
not really important

s1monw commented on src/main/java/org/elasticsearch/script/ScriptedOp.java:
```diff
@@ -0,0 +1,48 @@
+ * Operation/api that uses a script as part of its execution.
+ * Note that the suggest api is considered part of search for simplicity, as well as the percolate api.
+ */
+public enum ScriptedOp {
```
again do we have to have different / alternative names? this can just be strict no? just go with whatever we have though...

s1monw replied:
what is mapping here again? and don't we use it as well in the fetch phase so we should have one for this too?

javanna replied:
fetch falls under search, as well as percolator and suggester (this is documented). mapping covers the transfom mapping feature. The rationale is try not to have too many categories, I discussed this with @clintongormley but maybe we can change this.

The thing around alternate names is mainly that I know people are going to either use `mapping` or `mappings` and I want to be nice to them since it has the same meaning :) same for aggs and aggregations, those are synonyms as far as I'm concerned

s1monw replied:
don't add so many options one for each is fine nobody needs `mappings` vs `mapping` we can barf if it's wrong and provide a useful errormessage. please remove all the alternative names.

javanna replied:
There might be other settings that start with "script.". I don't think we can just barf whenever we find something that we don't recognize from within `ScriptModes`. I am ok with removing alternative names but we will end up ignoring what we don't know about I am afraid.

s1monw commented on src/main/java/org/elasticsearch/script/ScriptedOp.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Operation/api that uses a script as part of its execution.
+ * Note that the suggest api is considered part of search for simplicity, as well as the percolate api.
+ */
```
I really don't like this name, I think `ExecutionContext` or `ScriptContext` might be better names?

javanna replied:
dunno, I think *Context is too generic and we use it in other places with different meanings. To me this is really the operation/api that is trying to execute a script, hence why I came up with `ScriptedOp`. That said I don't have a super strong opinion about it.

s1monw replied:
it is really the context under which the script is executed though IMO...

javanna replied:
renamed

s1monw commented:
Overall this looks very very good @javanna good job! I left a bunch of comments...

javanna commented:
I pushed a few more commits to address review comments, and replied to all of them. Sorry if the comments got messed up, I rebased by mistake, not intentionally...

javanna commented on src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -579,6 +609,15 @@ public static void writeTo(ScriptType scriptType, StreamOutput out) throws IOExc
+        public String toString() {
+            //ScriptModes looks for 'dynamic' rather than 'inline' when reading settings, we should rename but we want to keep bw comp for now.
+            if (this == INLINE) {
+                return "dynamic";
```
@clintongormley we use `inline` instead of `dynamic` to refer to dynamic scripts... I added this special case to convert ScriptType.INLINE to the `dynamic` string in a bw comp manner. At some point we will have to choose for one of the two. Wondering which one we prefer: either use inline in settings here or rename ScriptType.INLINE to DYNAMIC in the future (master, it's a breaking change).

clintongormley replied:
Honestly I prefer inline, but all the user-facing stuff has been called dynamic, so I'd probably stick with that.

javanna replied:
oh boy that means that we should keep this distinction in master too, cause it's not just about bw comp. we prefer inline internally but we always called it dynamic when talking to the outside world... but I see your point, coming up with inline now among settings is confusing.

clintongormley commented on docs/reference/modules/scripting.asciidoc:
```diff
@@ -225,39 +223,109 @@ curl -XDELETE localhost:9200/_scripts/groovy/indexedCalculateScore
+| `enable`  |scripting is turned on, in the context of the setting being set.
+| `sandbox` |scripts may be executed only for languages that are sandboxed.
+|=======================================================================
+
```
Possibly mention that only expressions is sandboxed?

javanna replied:
mustache is too :) I will look into clarifying

clintongormley commented:
Docs look great @javanna

javanna commented:
@s1monw can you have another look please?

s1monw commented on src/main/java/org/elasticsearch/script/ScriptModes.java:
```diff
@@ -0,0 +1,215 @@
+    final ImmutableMap<String, ScriptMode> scriptModes;
+
+    ScriptModes(Map<String, ScriptEngineService> scriptEngines, Settings settings) {
+        ESLogger logger = Loggers.getLogger(getClass(), settings);
```
this is not what I meant in my comment. I wanted to get rid of the logger in this class entirely. Make it a simple class where all it's calculations get consumed by another. That makes it easy to test though...

s1monw commented on src/main/java/org/elasticsearch/script/ScriptModes.java:
```diff
@@ -0,0 +1,215 @@
+                        break;
+                }
+            } else {
+                logger.warn("ignoring [{}] setting as conflicting scripting settings have been specified: {}", ScriptService.DISABLE_DYNAMIC_SCRIPTING_SETTING, processedSettings);
```
if there is a conflict we should really throw an exception instead of swallowing it?

javanna replied:
that would be an easy way to get rid of the logger too... :) I discussed this with @clintongormley and we agreed on only logging and ignoring old settings, but sure we can throw an exception and it will make the node startup fail.

s1monw replied:
ok fair enough. I think for this delicate issue we should force the right setting IMO

javanna replied:
agreed I changed it it throw exception instead

s1monw commented on src/main/java/org/elasticsearch/script/ScriptService.java:
```diff
@@ -579,6 +609,16 @@ public static void writeTo(ScriptType scriptType, StreamOutput out) throws IOExc
+        public String toString() {
+            //ScriptModes looks for 'dynamic' rather than 'inline' when reading settings, that's the terminology
+            //we always used with users, although we prefer inline internally
+            if (this == INLINE) {
```
this is very confusing.. please use INLINE though

javanna replied:
I see the confusion... have a look at this discussion: https://github.com/elastic/elasticsearch/pull/10116#discussion_r26744097 . @clintongormley @s1monw can you guys come to an agreement? :) 
Using inline would simplify things but I don't like the fact that we change terminology for our users to just simplify our code. I am more for renaming the ScriptType.INLINE to DYNAMIC later on on master and leave this as-is for now.

s1monw replied:
we are talking about  toString method?! No way we add a distinction in there... if we pass this to the user then fix it in the rest layer

uboness replied:
+1 on @s1monw's comment... though I'd still reconsider changing it in the rest layer... if we change it to the user, this change should be reflected in the code as well (e.g. java clients)... I'd keep it "inline" everywhere for now...

javanna replied:
it's not only a REST problem. we use this to parse the settings too. we convert the enum name to lowercase string (and to dynamic if inline). We might also print out settings to users in certain situations, in which case we do want to print out `dynamic` rather than `inline`. we could move this out of the toString but then we would have the same thing done twice, and the problem is still the same. To me this condition is there only to keep bw comp in 1.x, but we have to remove in master and make up our minds between `inline` and `dynamic`, which we cannot really do in 1.x as we have this terminology mismatch already (since we introduced INLINE instead of DYNAMIC a while ago).

s1monw replied:
ok let's drop it in master and keep it here for BWC

javanna replied:
been thinking about it... we might want to take the chance to introduce `inline` to users with this PR and remove this conversion. My thinking is that dy `dynamic` up until now we meant indexed scripts + inline scripts (script.disable_dynamic setting would affect both), while now we have a distinction between the two. dynamic in the context of fine-grained settings becomes even more confusing maybe, we should really use inline and indexed now.
Thoughts? Isn't it going to be better for users too? @clintongormley ?

s1monw replied:
I agree... lets do it!

s1monw commented:
left some more comments

javanna commented:
I pushed a few more commits that address the comments I got. The only item that is still under discussion is about the `inline` vs `dynamic` terminology above.

s1monw commented:
LGTM

javanna commented:
I updated again, to use `inline` everywhere instead of `dynamic`, @s1monw you may want to have a last look. I also adjusted the docs a bit (added coming tags and a note about mustache scripts and search templates). I think this is ready, pretty much waiting for @clintongormley 's blessing :)

clintongormley commented:
LGTM

javanna commented:
marking as breaking as it breaks plugins that make use of `ScriptService`, for instance rivers that might allow to modify documents via scripts before indexing them. It doesn't break lang plugins though. ing @dadoonet :)

javanna commented:
Removing the breaking label, this is not breaking anymore after #10419.
